parts/kubernetesmastercustomdata.yml

```diff
@@ -335,7 +338,57 @@ write_files:
     fi
     mount $MOUNTPOINT
 
+{{if .OrchestratorProfile.IsCustomEtcdVersion}}
+- path: "/etc/systemd/system/etcd.service"
+  permissions: "0644"
+  owner: "root"
+  content: |
+    [Unit]
+    Description=etcd - highly-available key value store
+    Documentation=https://github.com/coreos/etcd
+    Documentation=man:etcd
+    After=network.target
+    Wants=network-online.target
+    [Service]
+    Environment=DAEMON_ARGS=
+    Environment=ETCD_NAME=%H
+    Environment=ETCD_DATA_DIR=/var/lib/etcd/default
+    EnvironmentFile=-/etc/default/%p
+    Type=notify
+    User=etcd
+    PermissionsStartOnly=true
+    ExecStart=/usr/bin/etcd $DAEMON_ARGS
+    Restart=on-abnormal
+    [Install]
+    WantedBy=multi-user.target
+    Alias=etcd.service
+
+- path: "/opt/azure/containers/setup-etcd.sh"
+  permissions: "0744"
+  owner: "root"
+  content: |
+    #!/bin/bash
+    # if provisioning using any etcd version other than 2.5.2
+    # We download it and set it up here 
+    set -x
+    ETCD_VER=v{{ .OrchestratorProfile.KubernetesConfig.EtcdVersion }}
+    DOWNLOAD_URL=https://github.com/coreos/etcd/releases/download # we default to github
+    mkdir -p /tmp/etcd-download
+    curl -L ${DOWNLOAD_URL}/${ETCD_VER}/etcd-${ETCD_VER}-linux-amd64.tar.gz -o /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz
+    tar xzvf /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz -C /usr/bin/ --strip-components=1
+    #create etcd user
+    useradd -U "etcd"
+    usermod -p "$(head -c 32 /dev/urandom | base64)" "etcd"
+    passwd -u "$etcd" > /dev/null
+    # update systemctl and start service
+    systemctl daemon-reload 
+    systemctl enable etcd.service
+{{end}}
+
 runcmd: 
+{{ if .OrchestratorProfile.IsCustomEtcdVersion }}
+- /opt/azure/containers/setup-etcd.sh
+{{end}}
 - apt-mark hold walinuxagent {{GetKubernetesMasterPreprovisionYaml}}
 - /bin/echo DAEMON_ARGS=--name "{{WrapAsVerbatim "variables('masterVMNames')[copyIndex(variables('masterOffset'))]"}}" --initial-advertise-peer-urls "{{WrapAsVerbatim "variables('masterEtcdPeerURLs')[copyIndex(variables('masterOffset'))]"}}" --listen-peer-urls "{{WrapAsVerbatim "variables('masterEtcdPeerURLs')[copyIndex(variables('masterOffset'))]"}}" --advertise-client-urls "{{WrapAsVerbatim "variables('masterEtcdClientURLs')[copyIndex(variables('masterOffset'))]"}}" --listen-client-urls "{{WrapAsVerbatim "concat(variables('masterEtcdClientURLs')[copyIndex(variables('masterOffset'))], ',http://127.0.0.1:', variables('masterEtcdClientPort'))"}}" --initial-cluster-token "k8s-etcd-cluster" --initial-cluster "{{WrapAsVerbatim "variables('masterEtcdClusterStates')[div(variables('masterCount'), 2)]"}} --data-dir "/var/lib/etcddisk"" --initial-cluster-state "new" | tee -a /etc/default/etcd
 - sudo /bin/chown -R etcd:etcd /var/lib/etcd/default
```

Comment:
Does this still work? from line 394-401.  You probably have to guard the custom configuration of etcd.